### ISSUE TITLE
Web serving database access denied

### ISSUE THREAD

**Issue by ahmadelyoussef** [parsa-epfl/cloudsuite#449]
I have started the database, memcached server and the web server containers successfully. When I try to curl -L http://localhost:8080, I'm getting the following error message in /varl/log/nginx/elgg_error.log
An exception occurred in the driver: SQLSTATE[HY000] [1698] Access denied for user 'root'@'localhost'.

How can I resolve this issue?

**Comment by toffernator** [comment]
TLDR:

The problem is with the configuration of `cloudsuite/web-serving:db_server`.
It does not allow the `root` database user to login from `localhost`.
You can allow this by executing SQL directly on the database_server container:
```bash
docker exec -it database_server mysql -u root --password=root -e "GRANT ALL PRIVILEGES ON ELGG_DB.* TO 'root'@'localhost' IDENTIFIED BY 'root'; FLUSH PRIVILEGES;"
```

---

I stumbled this same issue. I've run the database, memcache server, and the webserver according to https://github.com/parsa-epfl/cloudsuite/blob/main/docs/benchmarks/web-serving.md. For clarity, I use the commands:
```bash
docker run -it --net=host --name=database_server cloudsuite/web-serving:db_server
```
```bash
docker run -dt --net=host --name=memcache_server cloudsuite/web-serving:memcached_server
```
I wait for the database_server to have fully downloaded the data before starting the web server:
```bash
docker run -dt --net=host --name=web_server cloudsuite/web-serving:web_server /etc/bootstrap.sh http 127.0.0.1 127.0.0.1 127.0.0.1 4 auto
```

When I `curl locahost:8080` (or `curl localhost:8443` when the protocol is set to https for the web_server) I get the response:
```
Fatal error in exception handler. Check log for Exception at time 1739462105
```

When I check the web_server's logs (`docker exec -it web_server tail /var/log/nginx/elgg_error.log`) I observe the following which correlates with the Exception time:
```
An exception occurred in the driver: SQLSTATE[HY000] [1698] Access denied for user 'root'@'localhost'
  {"backtrace":["[#10] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Database/Select.php:14","[#9] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Database/ConfigTable.php:145","[#8] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/BootService.php:48","[#7] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Application/BootHandler.php:78","[#6] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Application/BootHandler.php:39","[#5] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Application.php:189","[#4] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Application.php:449","[#3] /usr/share/nginx/html/elgg/vendor/elgg/elgg/engine/classes/Elgg/Application.php:345","[#2] /usr/share/nginx/html/elgg/vendor/elgg/elgg...PHP message: [2025-02-13T15:55:05+00:00] ELGG.CRITICAL: Exception at time 1739462105:
Elgg couldn't connect to the database using the given credentials. Check the settings file.
```

I assume that the "settings file" I need to check is `/usr/share/nginx/html/elgg/elgg-config/settings.php` (`docker exec -it web_server cat /usr/share/nginx/html/elgg/elgg-config/settings.php`. In this file I see can that `dbuser` and `dbpass` are set to `root`. As an aside, the `dbname` is set to `ELGG_DB`:

```php
// ...

/**
 * The database username
 *
 * @global string $CONFIG->dbuser
 */
$CONFIG->dbuser = 'root';

/**
 * The database password
 *
 * @global string $CONFIG->dbpass
 */
$CONFIG->dbpass = 'root';

/**
 * The database name
 *
 * @global string $CONFIG->dbname
 */
$CONFIG->dbname = 'ELGG_DB';

// ...
```

I can login to the mariadb directly from the db_server container:
```bash
docker exec -it database_server mariadb --user=root --password=root ELGG_DB
```

So it's likely an issue about how authentication the database_server is configured. 
I configure mariadb to allow the `root` user to login from `localhost` with the following command:
```bash
docker exec -it database_server mysql -u root --password=root -e "GRANT ALL PRIVILEGES ON ELGG_DB.* TO 'root'@'localhost' IDENTIFIED BY 'root'; FLUSH PRIVILEGES;"
```

Now, when I `curl localhost:8080` I get back a web-page:
```
...
var elgg = {"config":{"lastcache":1660314795,"viewtype":"default","simplecache_enabled":1,"current_language":"en"},"security":{"token":{"__elgg_ts":1739462888,"__elgg_token":"GzqZVr58N-FrIvUdQeFtGg"}},"session":{"user":null,"token":"W4J7t4qTAIfUozh_RBZYUA"},"_data":{}};
</script><script src="http://127.0.0.1:8080/cache/1660314795/default/jquery.js"></script><script src="http://127.0.0.1:8080/cache/1660314795/default/require.js"></script><script src="http://127.0.0.1:8080/cache/1660314795/default/elgg.js"></script><script>
require([
    "input/form",
    "icon/user/default",
    "elgg/likes",
    "page/elements/topbar",
    "elgg/popup"
]);
</script>
</body></html
```
